File: python/dsh_cae/cfd_mesh.py
"""CFD mesh stage: lay a parametric blockMesh case and check it. Geometry arrives
in mm and is converted to SI here — the only unit conversion in the CFD chain."""
import argparse
import pathlib
import subprocess

from dsh_cae import cfd_case
from dsh_cae.receipt import emit, fail


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--workdir", required=True)
    parser.add_argument("--name", default="duct")
    parser.add_argument("--length-mm", type=float, required=True)
    parser.add_argument("--width-mm", type=float, required=True)
    parser.add_argument("--height-mm", type=float, required=True)
    parser.add_argument("--cell-size-mm", type=float, required=True)
    parser.add_argument("--wall-grading", type=float, default=1.0)
    parser.add_argument("--block-mesh-dict-file")
    parser.add_argument("--bashrc")
    parser.add_argument("--timeout-s", type=int, default=600)
    args = parser.parse_args()

    if args.name in (".", "..") or "/" in args.name or "\\" in args.name:
        fail(f"case name '{args.name}' must be a plain directory name")
    for flag in ("length_mm", "width_mm", "height_mm", "cell_size_mm"):
        value = getattr(args, flag)
        if value <= 0:
            fail(f"--{flag.replace('_', '-')} must be positive, got {value}")
    bashrc = cfd_case.resolve_bashrc(args.bashrc)
    case_dir = pathlib.Path(args.workdir) / "cfd" / args.name
    case_dir.mkdir(parents=True, exist_ok=True)

    if args.block_mesh_dict_file:
        dict_text = pathlib.Path(args.block_mesh_dict_file).read_text()
    else:
        dict_text = cfd_case.render_block_mesh_dict(
            args.length_mm / 1000.0, args.width_mm / 1000.0,
            args.height_mm / 1000.0, args.cell_size_mm / 1000.0, args.wall_grading)
    cfd_case.write_case_skeleton(case_dir, dict_text)

    try:
        meshed = cfd_case.foam_run(bashrc, "blockMesh", ["-case", str(case_dir)], args.timeout_s)
    except subprocess.TimeoutExpired:
        fail(f"blockMesh timed out after {args.timeout_s}s")
    (case_dir / "blockMesh.log").write_text(meshed.stdout + meshed.stderr)
    if meshed.returncode != 0:
        fail(f"blockMesh failed with exit {meshed.returncode}:\n"
             f"{cfd_case.tail40(meshed.stdout + meshed.stderr)}")

    try:
        checked = cfd_case.foam_run(bashrc, "checkMesh", ["-case", str(case_dir)], args.timeout_s)
    except subprocess.TimeoutExpired:
        fail(f"checkMesh timed out after {args.timeout_s}s")
    check_log = case_dir / "checkMesh.log"
    check_log.write_text(checked.stdout + checked.stderr)

    metrics = cfd_case.parse_check_mesh(checked.stdout)
    emit({
        "caseDir": str(case_dir),
        "blockMeshDictPath": str(case_dir / "system" / "blockMeshDict"),
        "boundsM": metrics["boundsM"],
        "cells": metrics["cells"],
        "maxNonOrthogonalityDeg": metrics["maxNonOrthogonalityDeg"],
        "maxAspectRatio": metrics["maxAspectRatio"],
        "checksPassed": metrics["checksPassed"],
        "checkMeshLogPath": str(check_log),
        "logTail": cfd_case.tail40(checked.stdout),
    })


if __name__ == "__main__":
    main()

File: src/tools/cfd-mesh.ts
import { join, resolve } from 'node:path'
import { mkdir, writeFile } from 'node:fs/promises'
import { defineTool } from '@deepseek-ai/dsh-tools'
import { ensureDeps, runStage } from '../runner.js'
import type { Config } from '../config.js'

const DESCRIPTION =
  'Generate a CFD mesh for a rectangular duct with blockMesh and validate it with checkMesh. '
  + 'Geometry in mm; the chain converts to SI meters here and stays SI afterwards. Flow along +x, '
  + 'inlet at x-min; patches are named inlet/outlet/walls. Returns caseDir (pass to cae_cfd_steady), '
  + 'cell count, and mesh quality. checksPassed=false is a mesh-quality verdict, not an error: '
  + 'refine (smaller cellSizeMm) or use wallGrading, or supply blockMeshDict — full dict text '
  + '(convertToMeters 1, patches must still be named inlet/outlet/walls) — for anything '
  + 'non-rectangular. checksPassed=null means the checkMesh output could not be parsed: read '
  + 'checkMeshLogPath yourself.'

/** Receipt shape of the `cfd_mesh` stage, pinned for the tool's output schema. */
export interface CfdMeshReceipt {
  caseDir: string
  blockMeshDictPath: string
  boundsM: { min: number[]; max: number[] } | null
  cells: number | null
  maxNonOrthogonalityDeg: number | null
  maxAspectRatio: number | null
  checksPassed: boolean | null
  checkMeshLogPath: string
  logTail: string
}

/** Build the `cae_cfd_mesh` tool bound to one deployment configuration. */
export function defineCaeCfdMeshTool(config: Config) {
  return defineTool({
    name: 'cae_cfd_mesh',
    description: DESCRIPTION,
    parameters: {
      lengthMm: { type: 'number', required: true, description: 'Duct length in mm (flow direction x).' },
      widthMm: { type: 'number', required: true, description: 'Cross-section width in mm (y).' },
      heightMm: { type: 'number', required: true, description: 'Cross-section height in mm (z).' },
      cellSizeMm: { type: 'number', required: true, description: 'Target cell edge in mm; per-direction counts are rounded.' },
      wallGrading: { type: 'number', description: 'Symmetric center-to-wall expansion ratio in y/z (1 = uniform). Default 1.' },
      blockMeshDict: { type: 'string', description: 'Full blockMeshDict text (escape hatch); overrides the parametric geometry.' },
      name: { type: 'string', description: 'Case directory name under <workdir>/cfd/. Default "duct".' },
    },
    output: {
      schema: {
        type: 'object', additionalProperties: false,
        properties: {
          caseDir: { type: 'string', required: true },
          blockMeshDictPath: { type: 'string', required: true },
          boundsM: {
            oneOf: [{ type: 'object', additionalProperties: false, properties: {
              min: { type: 'array', items: { type: 'number' }, required: true },
              max: { type: 'array', items: { type: 'number' }, required: true },
            } }, { type: 'null' }], required: true,
          },
          cells: { oneOf: [{ type: 'integer' }, { type: 'null' }], required: true },
          maxNonOrthogonalityDeg: { oneOf: [{ type: 'number' }, { type: 'null' }], required: true },
          maxAspectRatio: { oneOf: [{ type: 'number' }, { type: 'null' }], required: true },
          checksPassed: { oneOf: [{ type: 'boolean' }, { type: 'null' }], required: true },
          checkMeshLogPath: { type: 'string', required: true },
          logTail: { type: 'string', required: true },
        },
      },
      render: (_args, value) => [{
        type: 'text',
        text: `blockMesh: ${value.cells ?? '?'} cells, max non-orthogonality `
          + `${value.maxNonOrthogonalityDeg ?? 'n/a'}°, max aspect ${value.maxAspectRatio ?? 'n/a'}, checks `
          + `${value.checksPassed === null ? 'unparsed — read checkMeshLogPath' : value.checksPassed ? 'OK' : 'FAILED (refine or escape via blockMeshDict)'}. `
          + `Case: ${value.caseDir}`,
      }],
    },
    async execute(args, exec) {
      const name = args.name ?? 'duct'
      if (name.includes('/') || name.includes('\\') || name === '.' || name === '..') {
        throw new Error(`case name '${name}' must be a plain directory name`)
      }
      await ensureDeps(config, exec.signal, 'cfd')
      const workdirAbs = resolve(config.workdir)
      await mkdir(join(workdirAbs, 'cfd'), { recursive: true })
      const argv = [
        '--workdir', workdirAbs,
        '--name', name,
        '--length-mm', String(args.lengthMm),
        '--width-mm', String(args.widthMm),
        '--height-mm', String(args.heightMm),
        '--cell-size-mm', String(args.cellSizeMm),
        '--wall-grading', String(args.wallGrading ?? 1),
      ]
      if (args.blockMeshDict) {
        const dictFile = join(workdirAbs, 'cfd', `${name}.blockMeshDict.txt`)
        await writeFile(dictFile, args.blockMeshDict, 'utf8')
        argv.push('--block-mesh-dict-file', dictFile)
      }
      if (config.openfoamBashrc) argv.push('--bashrc', config.openfoamBashrc)
      const { receipt } = await runStage(config, 'cfd_mesh', argv,
        { signal: exec.signal, logFile: `cfd.${name}.mesh.log` })
      return receipt as unknown as CfdMeshReceipt
    },
    presentCall: args => ({ card: 'terminal', title: `blockMesh ${args.name ?? 'duct'}`, description: 'CFD duct mesh' }),
  })
}
